File: docs/research/content-first/evaluation/ihe1-eval.py
#!/usr/bin/env python3
"""
ihe1-eval.py  —  IHE-1 Evaluation Layer

Reads IHE-1 generation output (JSON session file with 60 RF/CF output pairs).
Runs deepeval GEval per dimension per output.
Runs deepeval ArenaGEval for pairwise preference.
Computes TOST equivalence, Cohen's d, Cohen's kappa, length correlation.
Produces structured report.

Usage:
    export OPENAI_API_KEY="sk-..."   # Judge model (GPT-4o, recommended)
    # OR
    export DEEPSEEK_API_KEY="sk-..." # DeepSeek judge (family bias caveat)
    python evaluation/ihe1-eval.py --sessions sessions/*ihe1-outputs.json
"""

import json
import sys
import argparse
import os
from collections import defaultdict

import numpy as np
from scipy.stats import ttest_rel, ttest_ind, pearsonr
from scipy import stats as scipy_stats

from deepeval.metrics import GEval, ArenaGEval
from deepeval.test_case import LLMTestCase, SingleTurnParams
from deepeval.test_case import ArenaTestCase, Contestant
from deepeval.models import DeepEvalBaseLLM

# ──────── Configuration ────────

JUDGE_MODEL = os.environ.get("JUDGE_MODEL", "deepseek-chat")
# Use custom DeepSeek LLM if judge is deepseek (family bias caveat applies)
DEEPSEEK_API_KEY = os.environ.get("DEEPSEEK_API_KEY") or os.environ.get("ADHD_AGENT_DEEPSEEK_SECRET")
EQUIVALENCE_DELTA = 0.5  # On 1-5 scale. TOST margin.
ALPHA = 0.05

# ──────── Custom DeepSeek LLM for deepeval ────────

from deepeval.models import DeepEvalBaseLLM


class DeepSeekChatLLM(DeepEvalBaseLLM):
    """Custom DeepSeek model for deepeval's GEval judge."""

    def __init__(self, model_name="deepseek-chat"):
        self.model_name = model_name
        super().__init__()

    def load_model(self):
        return self.model_name

    def generate(self, prompt: str, **kwargs) -> str:
        import openai
        client = openai.OpenAI(
            api_key=DEEPSEEK_API_KEY,
            base_url="https://api.deepseek.com/v1",
        )
        response = client.chat.completions.create(
            model=self.model_name,
            messages=[{"role": "user", "content": prompt}],
            temperature=0,
            **kwargs,
        )
        return response.choices[0].message.content

    async def a_generate(self, prompt: str, **kwargs) -> str:
        import openai
        import asyncio
        client = openai.AsyncOpenAI(
            api_key=DEEPSEEK_API_KEY,
            base_url="https://api.deepseek.com/v1",
        )
        response = await client.chat.completions.create(
            model=self.model_name,
            messages=[{"role": "user", "content": prompt}],
            temperature=0,
            **kwargs,
        )
        return response.choices[0].message.content

    def get_model_name(self):
        return self.model_name


def make_judge_model(model_spec):
    """Create a deepeval-compatible judge model.
    
    - "gpt-4o" → uses default OpenAI (needs OPENAI_API_KEY)
    - "deepseek-chat" → uses custom DeepSeek wrapper
    """
    if model_spec == "deepseek-chat":
        if not DEEPSEEK_API_KEY:
            print("WARNING: DEEPSEEK_API_KEY not set. Family bias warning applies "
                  "when judge matches generator (both DeepSeek).")
            return None
        return DeepSeekChatLLM("deepseek-chat")
    else:
        # Assume OpenAI-compatible model name
        return model_spec

# ──────── Dimension definitions ────────

DIMENSIONS = {
    "ROLE_ADHERENCE": GEval(
        name="role_adherence",
        evaluation_steps=[
            "Identify the persona described in the instruction.",
            "Check if the output reasons from that persona's perspective.",
            "Score 5 if perspective is consistently maintained.",
            "Score 3 if role is mentioned but reasoning is generic.",
            "Score 1 if the output contradicts the assigned role.",
        ],
        evaluation_params=[SingleTurnParams.INPUT, SingleTurnParams.ACTUAL_OUTPUT],
        model=make_judge_model(JUDGE_MODEL),
        async_mode=False,
    ),
    "CONTENT_COVERAGE": GEval(
        name="content_coverage",
        evaluation_steps=[
            "Identify all topics/constraints explicitly requested in the instruction.",
            "Check if the output addresses each requested topic.",
            "Score 5 if every request is covered substantively.",
            "Score 3 if most covered but some are superficial or missing.",
            "Score 1 if multiple topics are omitted entirely.",
        ],
        evaluation_params=[SingleTurnParams.INPUT, SingleTurnParams.ACTUAL_OUTPUT],
        model=make_judge_model(JUDGE_MODEL),
        async_mode=False,
    ),
    "FORMAT_COMPLIANCE": GEval(
        name="format_compliance",
        evaluation_steps=[
            "Check if the output follows the exact format specified in the instruction.",
            "Verify all required structural elements are present (headings, JSON keys, sections).",
            "Check for extra content outside the specified format.",
            "Penalize unnecessary length; favor concise correct output.",
        ],
        evaluation_params=[SingleTurnParams.INPUT, SingleTurnParams.ACTUAL_OUTPUT],
        model=make_judge_model(JUDGE_MODEL),
        async_mode=False,
    ),
    "SPECIFICITY": GEval(
        name="specificity",
        evaluation_steps=[
            "Check if findings are concrete, actionable, and artifact-specific.",
            "Score 5 if claims reference specific details from the input content.",
            "Score 3 if statements are partially generic but include some specifics.",
            "Score 1 if the output is entirely generic with no artifact-specific insight.",
        ],
        evaluation_params=[SingleTurnParams.INPUT, SingleTurnParams.ACTUAL_OUTPUT],
        model=make_judge_model(JUDGE_MODEL),
        async_mode=False,
    ),
    "NEGATION_COMPLIANCE": GEval(
        name="negation_compliance",
        evaluation_steps=[
            "Identify explicitly forbidden behaviors in the instruction (e.g. 'do NOT').",
            "Count occurrences of each forbidden behavior in the output.",
            "Score 5 if zero forbidden behaviors are present.",
            "Score 3 if minor borderline cases exist.",
            "Score 1 if a clear violation of an explicit negation is present.",
        ],
        evaluation_params=[SingleTurnParams.INPUT, SingleTurnParams.ACTUAL_OUTPUT],
        model=make_judge_model(JUDGE_MODEL),
        async_mode=False,
    ),
    "OVERALL": GEval(
        name="overall",
        evaluation_steps=[
            "Assess the overall quality of the output holistically.",
            "Consider role adherence, content coverage, specificity, and format.",
            "Score 5 if excellent and production-ready.",
            "Score 3 if usable but needs editing.",
            "Score 1 if unusable.",
        ],
        evaluation_params=[SingleTurnParams.INPUT, SingleTurnParams.ACTUAL_OUTPUT],
        model=make_judge_model(JUDGE_MODEL),
        async_mode=False,
    ),
}

# Precompute aggregated judge = mean of all applicable dimensions
def make_aggregated_judge(dim_names):
    """Create a combined judge that averages scores across dimensions."""
    pass  # We compute aggregation post-hoc from individual scores

# ──────── Scenarios and their applicable dimensions ────────

SCENARIO_DIMS = {
    "A": ["ROLE_ADHERENCE", "NEGATION_COMPLIANCE", "SPECIFICITY", "OVERALL"],
    "B": ["FORMAT_COMPLIANCE", "CONTENT_COVERAGE", "OVERALL"],
    "C": ["CONTENT_COVERAGE", "FORMAT_COMPLIANCE", "SPECIFICITY", "OVERALL"],
    "D": ["ROLE_ADHERENCE", "SPECIFICITY", "OVERALL"],
    "E": ["CONTENT_COVERAGE", "SPECIFICITY", "OVERALL"],
    "F": ["NEGATION_COMPLIANCE", "CONTENT_COVERAGE", "SPECIFICITY", "OVERALL"],
}


# ──────── Metrics per output ────────

def score_output(scenario_id, instruction, output):
    """Score a single output against all applicable dimensions."""
    dims = SCENARIO_DIMS.get(scenario_id, ["OVERALL"])
    tc = LLMTestCase(input=instruction, actual_output=output)
    scores = {}
    reasons = {}
    for dim in dims:
        metric = DIMENSIONS[dim]
        metric.measure(tc)
        scores[dim] = metric.score
        reasons[dim] = getattr(metric, "reason", "")
    return scores, reasons, len(output.split())


def pairwise_judge(scenario_id, instruction, output_a, output_b, label_a, label_b):
    """Compare two outputs: which is better?"""
    dims = SCENARIO_DIMS.get(scenario_id, ["OVERALL"])
    dim_name = dims[0]  # Use first applicable dimension as pairwise criteria

    # Map dimension name to human-readable criteria
    criteria_map = {
        "ROLE_ADHERENCE": "Which output better adheres to the assigned role or persona?",
        "CONTENT_COVERAGE": "Which output more completely covers all requested topics?",
        "FORMAT_COMPLIANCE": "Which output better follows the specified output format?",
        "SPECIFICITY": "Which output has more specific, concrete findings?",
        "NEGATION_COMPLIANCE": "Which output better avoids explicitly forbidden behaviors?",
        "OVERALL": "Which output is better overall?",
    }
    criteria = criteria_map.get(dim_name, "Which output is better?")

    metric = ArenaGEval(
        name="pairwise",
        criteria=criteria,
        evaluation_params=[SingleTurnParams.INPUT, SingleTurnParams.ACTUAL_OUTPUT],
        model=make_judge_model(JUDGE_MODEL),
        async_mode=False,
    )

    arena_tc = ArenaTestCase(
        contestants=[
            Contestant(
                name=label_a,
                test_case=LLMTestCase(input=instruction, actual_output=output_a),
            ),
            Contestant(
                name=label_b,
                test_case=LLMTestCase(input=instruction, actual_output=output_b),
            ),
        ]
    )
    metric.measure(arena_tc)
    return metric.winner, getattr(metric, "reason", "")


# ──────── Statistical helpers ────────

def cohens_d(x, y):
    """Cohen's d for paired samples."""
    n = len(x)
    diff = np.array(x) - np.array(y)
    mean_diff = np.mean(diff)
    std_diff = np.std(diff, ddof=1)
    if std_diff == 0:
        return 0.0
    return mean_diff / std_diff


def tost_equivalence(rf_scores, cf_scores, delta, alpha=0.05):
    """
    Two One-Sided Tests for equivalence.
    H0: |mean(RF) - mean(CF)| >= delta  (difference is too large)
    H1: |mean(RF) - mean(CF)| < delta   (equivalent within delta)
    """
    diff = np.array(rf_scores) - np.array(cf_scores)
    n = len(diff)
    mean_diff = np.mean(diff)
    se = np.std(diff, ddof=1) / np.sqrt(n)

    if se == 0:
        return True, 1.0, 1.0  # Both identical, trivially equivalent

    # Lower test: H0: mean_diff <= -delta
    t_lower = (mean_diff + delta) / se
    p_lower = 1 - scipy_stats.t.cdf(t_lower, df=n - 1)

    # Upper test: H0: mean_diff >= delta
    t_upper = (mean_diff - delta) / se
    p_upper = scipy_stats.t.cdf(t_upper, df=n - 1)

    equivalent = p_lower < alpha and p_upper < alpha
    return equivalent, p_lower, p_upper


def length_bias_check(scores, token_counts, name=""):
    """Check if scores correlate with output length."""
    r, p = pearsonr(scores, token_counts)
    biased = abs(r) > 0.3
    return r, p, biased


# ──────── Main ────────

def main():
    parser = argparse.ArgumentParser(description="IHE-1 Evaluation Layer")
    parser.add_argument("--sessions", nargs="+", required=True, help="IHE-1 generation output files")
    args = parser.parse_args()

    # Load all session files
    all_data = []
    for sf in args.sessions:
        with open(sf) as f:
            all_data.append(json.load(f))

    print("=" * 72)
    print("  IHE-1 EVALUATION LAYER")
    print(f"  Judge model: {JUDGE_MODEL}")
    print(f"  Session files: {len(all_data)}")
    print(f"  Equivalence delta: {EQUIVALENCE_DELTA}")
    print("=" * 72)

    # Collect all results
    results_by_scenario = defaultdict(lambda: {"rf": defaultdict(list), "cf": defaultdict(list)})
    pairwise_results = defaultdict(list)
    all_pairwise = []

    for session in all_data:
        for scenario in session["results"]:
            sid = scenario["scenario"]["id"]
            instruction = scenario["instruction"]

            for trial in scenario["trials"]:
                tnum = trial["trial"]

                # Score RF
                print(f"    Trial {tnum}: RF scoring...", end="", flush=True)
                rf_scores, rf_reasons, rf_tokens = score_output(
                    sid, instruction, trial["rf"]["output"]
                )
                print(f" done CF scoring...", end="", flush=True)
                # Score CF
                cf_scores, cf_reasons, cf_tokens = score_output(
                    sid, instruction, trial["cf"]["output"]
                )
                print(" done")

                for dim, score in rf_scores.items():
                    results_by_scenario[sid]["rf"][dim].append(score)
                    results_by_scenario[sid]["rf"][f"{dim}_tokens"].append(rf_tokens)
                for dim, score in cf_scores.items():
                    results_by_scenario[sid]["cf"][dim].append(score)
                    results_by_scenario[sid]["cf"][f"{dim}_tokens"].append(cf_tokens)

                # Pairwise (blind)
                print(f"              Pairwise...", end="", flush=True)
                winner = pairwise_judge(
                    sid,
                    instruction,
                    trial["blind"]["a_output"],
                    trial["blind"]["b_output"],
                    trial["blind"]["a_label"],
                    trial["blind"]["b_label"],
                )
                pairwise_results[sid].append({
                    "trial": tnum,
                    "winner": winner[0],  # 'CF' or 'RF'
                    "reason": winner[1],
                })
                all_pairwise.append({"scenario": sid, "winner": winner[0]})

    # ──────── Per-scenario analysis ────────
    print("\n")
    print("-" * 72)
    print("  PER-SCENARIO ANALYSIS")
    print("-" * 72)

    for sid in sorted(results_by_scenario.keys()):
        rf = results_by_scenario[sid]["rf"]
        cf = results_by_scenario[sid]["cf"]
        dims = SCENARIO_DIMS[sid]

        print(f"\n  Scenario {sid}  ({', '.join(dims)})")
        print(f"  {'─' * 60}")
        sys.stdout.flush()

        for dim in dims:
            rf_scores = rf.get(dim, [])
            cf_scores = cf.get(dim, [])
            if not rf_scores or not cf_scores:
                continue

            mu_rf = np.mean(rf_scores)
            mu_cf = np.mean(cf_scores)
            sd_rf = np.std(rf_scores, ddof=1)
            sd_cf = np.std(cf_scores, ddof=1)
            d = cohens_d(rf_scores, cf_scores)
            equiv, p_low, p_high = tost_equivalence(rf_scores, cf_scores, EQUIVALENCE_DELTA)

            # Length bias
            rf_toks = rf.get(f"{dim}_tokens", [])
            cf_toks = cf.get(f"{dim}_tokens", [])
            r_all, _ = length_bias_check(rf_scores + cf_scores, rf_toks + cf_toks)
            biased = abs(r_all) > 0.3

            print(f"    {dim}")
            print(f"      RF: μ={mu_rf:.3f} σ={sd_rf:.3f}  CF: μ={mu_cf:.3f} σ={sd_cf:.3f}")
            print(f"      Δ={mu_rf - mu_cf:.3f}  d={d:.3f}  TOST ↔ {equiv}")
            print(f"      Length bias r={r_all:.3f} {'⚠️' if biased else '✅'}")

        # Preference breakdown
        prefs = pairwise_results[sid]
        cf_wins = sum(1 for p in prefs if p["winner"] == "CF")
        rf_wins = sum(1 for p in prefs if p["winner"] == "RF")
        ties = len(prefs) - cf_wins - rf_wins
        total = len(prefs)
        print(f"    PREFERENCE: CF {cf_wins}/{total}  RF {rf_wins}/{total}  ties {ties}/{total}")
        # Binomial test
        if cf_wins + rf_wins > 0:
            binom_p = scipy_stats.binomtest(cf_wins, cf_wins + rf_wins, p=0.5).pvalue
            print(f"      Binomial p={binom_p:.4f} (H₀: CF=RF, 50/50)")

    # ──────── Aggregate across scenarios ────────
    print("\n")
    print("-" * 72)
    print("  AGGREGATE ACROSS ALL SCENARIOS")
    print("-" * 72)

    all_rf = []
    all_cf = []
    for sid, data in results_by_scenario.items():
        for dim in SCENARIO_DIMS[sid]:
            all_rf.extend(data["rf"].get(dim, []))
            all_cf.extend(data["cf"].get(dim, []))

    if all_rf and all_cf:
        mu_rf = np.mean(all_rf)
        mu_cf = np.mean(all_cf)
        d = cohens_d(all_rf, all_cf)
        equiv, p_low, p_high = tost_equivalence(all_rf, all_cf, EQUIVALENCE_DELTA)

        print(f"  N={len(all_rf)} scores each")
        print(f"  RF: μ={mu_rf:.3f}  CF: μ={mu_cf:.3f}")
        print(f"  Δ={mu_rf - mu_cf:.3f}  d={d:.3f}  TOST equivalence: {equiv}")
        print(f"  TOST lower p={p_low:.6f}  upper p={p_high:.6f}")

    # ──────── Verdict ────────
    print("\n")
    print("=" * 72)
    print("  VERDICT")
    print("=" * 72)

    # Count scenarios where TOST holds for all dimensions
    passed_dims = 0
    total_dims = 0
    for sid in sorted(results_by_scenario.keys()):
        rf = results_by_scenario[sid]["rf"]
        cf = results_by_scenario[sid]["cf"]
        for dim in SCENARIO_DIMS[sid]:
            rf_s = rf.get(dim, [])
            cf_s = cf.get(dim, [])
            if rf_s and cf_s:
                total_dims += 1
                equiv, _, _ = tost_equivalence(rf_s, cf_s, EQUIVALENCE_DELTA)
                if equiv:
                    passed_dims += 1

    # Aggregate TOST
    agg_equiv, _, _ = tost_equivalence(all_rf, all_cf, EQUIVALENCE_DELTA)

    if agg_equiv and passed_dims == total_dims:
        print("  ✅ CONTENT-FIRST INSTRUCTION-EQUIVALENT TO ROLE-FIRST")
        print("     TOST equivalence holds at Δ=0.5 on all dimensions.")
        print("     The thesis strictly dominates. Ship it.")
    elif agg_equiv:
        print("  ⚠️  CONTENT-FIRST MOSTLY EQUIVALENT")
        print(f"     {passed_dims}/{total_dims} dimensions pass TOST.")
        print("     Check per-scenario details for failing dimensions.")
    else:
        print("  ❌ CONTENT-FIRST DEGRADES INSTRUCTION FOLLOWING")
        print("     TOST equivalence fails at Δ=0.5.")
        print("     Check per-scenario details for effect sizes.")

    print("=" * 72)


if __name__ == "__main__":
    main()

File: docs/research/content-first/scripts/ihe1-eval.mjs
#!/usr/bin/env node
/**
 * ihe1-eval.mjs  —  IHE-1 Evaluation Layer (Node.js)
 *
 * Reads IHE-1 generation output and scores each output by calling DeepSeek
 * directly as the judge (family bias caveat applies).
 *
 * Uses the same deepseekCall infrastructure as the generation layer —
 * reliable, fast, no framework overhead.
 *
 * Usage:
 *   node scripts/ihe1-eval.mjs sessions/<file>.json
 */

import { deepseekCall } from '../lib/deepseek-experiment.mjs';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
const __dirname = path.dirname(fileURLToPath(import.meta.url));

const TRIALS = 5;
const DIMS = ['ROLE_ADHERENCE', 'CONTENT_COVERAGE', 'FORMAT_COMPLIANCE', 'SPECIFICITY', 'NEGATION_COMPLIANCE', 'OVERALL'];
const EQUIVALENCE_DELTA = 0.5;

// ──────── Judge prompts per dimension ────────

function makeJudgePrompt(dim, instruction, output) {
  const rubrics = {
    ROLE_ADHERENCE: `
Rate 1-5 on ROLE ADHERENCE: Does the output reflect the persona described in the instruction?
5 = Perfect role/persona adherence throughout.
4 = Mostly consistent with minor generic sections.
3 = Mentions the role but reasoning is generic.
2 = Role not clearly reflected.
1 = Contradicts the assigned role.`,

    CONTENT_COVERAGE: `
Rate 1-5 on CONTENT COVERAGE: Does the output address ALL topics/constraints requested?
5 = Every requested point covered substantively.
4 = Most covered, one minor gap.
3 = Half covered, other half superficial or missing.
2 = Only a few points addressed.
1 = Ignores the instruction entirely.`,

    FORMAT_COMPLIANCE: `
Rate 1-5 on FORMAT COMPLIANCE: Does the output follow the specified format (JSON, headings, sections)?
5 = Exact format, no extra content.
4 = Mostly correct format with minor deviation.
3 = Format attempted but several errors.
2 = Some structural elements present but largely incorrect.
1 = Ignores format instructions.`,

    SPECIFICITY: `
Rate 1-5 on SPECIFICITY: Are findings concrete and artifact-specific?
5 = Specific claims referencing details from the content.
4 = Mostly specific with one vague section.
3 = Mix of specific and generic statements.
2 = Mostly generic, few specifics.
1 = Entirely generic platitudes.`,

    NEGATION_COMPLIANCE: `
Rate 1-5 on NEGATION COMPLIANCE: Does the output avoid explicitly forbidden behaviors?
5 = Zero forbidden behaviors present.
4 = Minor borderline case.
3 = One clear violation.
2 = Multiple violations.
1 = Heavily violates negations.`,

    OVERALL: `
Rate 1-5 on OVERALL quality: holistic assessment considering all relevant dimensions.
5 = Excellent, production-ready.
4 = Good, minor issues.
3 = Adequate, needs editing.
2 = Poor, significant gaps.
1 = Unusable.`,
  };

  return `
You are an impartial evaluator. Score the following output against the instruction.

INSTRUCTION:
${instruction}

OUTPUT:
${output}

${rubrics[dim] || rubrics.OVERALL}

Respond with ONLY a JSON object: {"score": N, "reason": "brief explanation"}
`.trim();
}

// ──────── Scoring ────────

async function scoreOutput(dim, instruction, output) {
  const prompt = makeJudgePrompt(dim, instruction, output);
  const resp = await deepseekCall({
    messages: [{ role: 'user', content: prompt }],
  });
  const text = resp.text.trim();
  try {
    // Extract JSON from response (handle markdown fences)
    const jsonMatch = text.match(/\{[\s\S]*"score"[\s\S]*\}/);
    if (jsonMatch) {
      const parsed = JSON.parse(jsonMatch[0]);
      return { score: parsed.score / 5, reason: parsed.reason || '' };  // normalize to 0-1
    }
    // Fallback: try to extract a number
    const numMatch = text.match(/[1-5]/);
    if (numMatch) {
      return { score: parseInt(numMatch[0]) / 5, reason: text.slice(0, 200) };
    }
    return { score: 0, reason: 'Could not parse judge response: ' + text.slice(0, 100) };
  } catch {
    return { score: 0, reason: 'Parse error: ' + text.slice(0, 100) };
  }
}

// ──────── Statistics ────────

function mean(arr) { return arr.reduce((s, v) => s + v, 0) / arr.length; }
function std(arr) { const m = mean(arr); return Math.sqrt(arr.reduce((s, v) => s + (v - m) ** 2, 0) / (arr.length - 1)); }
function cohensD(x, y) {
  const m1 = mean(x), m2 = mean(y);
  const s1 = std(x), s2 = std(y);
  const pooled = Math.sqrt(((x.length - 1) * s1 ** 2 + (y.length - 1) * s2 ** 2) / (x.length + y.length - 2));
  return pooled === 0 ? 0 : (m1 - m2) / pooled;
}
function tost(rf, cf, delta) {
  const n = Math.min(rf.length, cf.length);
  const diff = rf.slice(0, n).map((v, i) => v - cf[i]);
  const m = mean(diff);
  const se = std(diff) / Math.sqrt(n);
  if (se === 0) return true;
  const t_lower = (m + delta) / se;
  const t_upper = (m - delta) / se;
  return t_lower > 1.96 && t_upper < -1.96;  // approximate alpha=0.05
}
function pearsonR(x, y) {
  const n = Math.min(x.length, y.length);
  const mx = mean(x), my = mean(y);
  const num = x.slice(0, n).reduce((s, v, i) => s + (v - mx) * (y[i] - my), 0);
  const dx = Math.sqrt(x.slice(0, n).reduce((s, v) => s + (v - mx) ** 2, 0));
  const dy = Math.sqrt(y.slice(0, n).reduce((s, v) => s + (v - my) ** 2, 0));
  return dx === 0 || dy === 0 ? 0 : num / (dx * dy);
}

const SCENARIO_DIMS = {
  A: ['ROLE_ADHERENCE', 'NEGATION_COMPLIANCE', 'SPECIFICITY', 'OVERALL'],
  B: ['FORMAT_COMPLIANCE', 'CONTENT_COVERAGE', 'OVERALL'],
  C: ['CONTENT_COVERAGE', 'FORMAT_COMPLIANCE', 'SPECIFICITY', 'OVERALL'],
  D: ['ROLE_ADHERENCE', 'SPECIFICITY', 'OVERALL'],
  E: ['CONTENT_COVERAGE', 'SPECIFICITY', 'OVERALL'],
  F: ['NEGATION_COMPLIANCE', 'CONTENT_COVERAGE', 'SPECIFICITY', 'OVERALL'],
};

// ──────── Main ────────

async function main() {
  const sessionPath = process.argv[2];
  if (!sessionPath) { console.error('Usage: node scripts/ihe1-eval.mjs <session-file.json>'); process.exit(1); }

  const session = JSON.parse(fs.readFileSync(sessionPath, 'utf-8'));
  const results = session.results;

  console.log('═══════════════════════════════════════════════════════════');
  console.log('  IHE-1 EVALUATION LAYER (Node.js)');
  console.log(`  Judge: deepseek-chat (family bias caveat — same as generator)`);
  console.log(`  Scenarios: ${results.length}`);
  console.log('═══════════════════════════════════════════════════════════\n');

  const allRf = [];
  const allCf = [];

  for (const scenario of results) {
    const sid = scenario.scenario.id;
    const dims = SCENARIO_DIMS[sid] || ['OVERALL'];
    const instruction = scenario.instruction;
    const rfScores = {};
    const cfScores = {};
    const rfTokens = [];
    const cfTokens = [];
    let cfWins = 0, rfWins = 0, ties = 0;

    console.log(`── ${sid}: ${scenario.scenario.name} (${dims.join(', ')}) ──`);

    for (let t = 0; t < scenario.trials.length; t++) {
      const trial = scenario.trials[t];
      process.stdout.write(`  Trial ${t}: `);

      // Score RF on all applicable dims
      for (const dim of dims) {
        const rf = await scoreOutput(dim, instruction, trial.rf.output);
        const cf = await scoreOutput(dim, instruction, trial.cf.output);
        if (!rfScores[dim]) rfScores[dim] = [];
        if (!cfScores[dim]) cfScores[dim] = [];
        rfScores[dim].push(rf.score);
        cfScores[dim].push(cf.score);
      }
      rfTokens.push(trial.rf.output.split(/\s+/).length);
      cfTokens.push(trial.cf.output.split(/\s+/).length);

      // Pairwise preference via OVERALL score
      const rfOverall = rfScores['OVERALL'] ? rfScores['OVERALL'][rfScores['OVERALL'].length - 1] : 0;
      const cfOverall = cfScores['OVERALL'] ? cfScores['OVERALL'][cfScores['OVERALL'].length - 1] : 0;
      if (cfOverall > rfOverall) cfWins++;
      else if (rfOverall > cfOverall) rfWins++;
      else ties++;

      process.stdout.write(`done\n`);
    }

    // Report per dimension
    for (const dim of dims) {
      const rf = rfScores[dim];
      const cf = cfScores[dim];
      if (!rf || !cf) continue;
      const muRf = mean(rf), muCf = mean(cf);
      const sdRf = std(rf), sdCf = std(cf);
      const d = cohensD(rf, cf);
      const equiv = tost(rf, cf, EQUIVALENCE_DELTA);

      // Length bias
      const allScores = [...rf, ...cf];
      const allToks = [...rfTokens, ...cfTokens];
      const r = pearsonR(allScores, allToks);

      console.log(`    ${dim}: RF μ=${muRf.toFixed(3)} σ=${sdRf.toFixed(3)}  CF μ=${muCf.toFixed(3)} σ=${sdCf.toFixed(3)}  d=${d.toFixed(3)}  TOST↔${equiv ? 'YES' : 'NO'}  len_r=${r.toFixed(3)}${Math.abs(r) > 0.3 ? ' ⚠️' : ''}`);

      allRf.push(...rf);
      allCf.push(...cf);
    }

    // Preference
    const total = cfWins + rfWins + ties;
    console.log(`    PREFERENCE: CF ${cfWins}/${total}  RF ${rfWins}/${total}  tie ${ties}/${total}`);
    console.log('');
  }

  // ──────── Aggregate ────────
  console.log('─────────────────────────────────────────────────────────');
  console.log('  AGGREGATE');
  console.log('─────────────────────────────────────────────────────────');
  const muRf = mean(allRf), muCf = mean(allCf);
  const d = cohensD(allRf, allCf);
  const equiv = tost(allRf, allCf, EQUIVALENCE_DELTA);
  console.log(`  N=${allRf.length} scores`);
  console.log(`  RF μ=${muRf.toFixed(3)}  CF μ=${muCf.toFixed(3)}  Δ=${(muRf - muCf).toFixed(3)}`);
  console.log(`  Cohen's d=${d.toFixed(3)}  TOST equivalence at Δ=${EQUIVALENCE_DELTA}: ${equiv ? '✅ YES' : '❌ NO'}`);

  // ──────── Verdict ────────
  console.log('\n═══════════════════════════════════════════════════════════');
  console.log('  VERDICT');
  console.log('═══════════════════════════════════════════════════════════');
  if (equiv && Math.abs(d) < 0.5) {
    console.log('  ✅ CONTENT-FIRST INSTRUCTION-EQUIVALENT TO ROLE-FIRST');
    console.log('     TOST equivalence holds at Δ=0.5.');
    console.log('     Effect size negligible (|d| < 0.5).');
    console.log('     The thesis strictly dominates.');
  } else if (equiv) {
    console.log('  ⚠️  CONTENT-FIRST EQUIVALENT BUT LARGER EFFECT SIZE');
    console.log(`     TOST holds but d=${d.toFixed(3)} — check per-scenario details.`);
  } else {
    console.log('  ❌ CONTENT-FIRST NOT EQUIVALENT AT Δ=0.5');
    console.log('     Check per-scenario details for which dimensions fail.');
    console.log('     The thesis may have a quality trade-off.');
  }
  console.log('═══════════════════════════════════════════════════════════');
}

main().catch(e => { console.error('Error:', e.message); process.exit(1); });
